### PR TITLE
added set functions (fix #15)

### LINKED ISSUE #15: export set functions

### PR DESCRIPTION
**Pull request by schochastics**

### CODE REVIEW

**Comment by schochastics** [comment]
Not sure what is going on with username and password
``` r
url <- "https://user_1:password_1@example.org:8080/dir/../api?q=1#frag"
adaR::ada_set_password(url, "hunter2")
#> [1] "https://user_1:hunter2"
adaR::ada_set_username(url, "user_2")
#> [1] "https://user_2"
```

File: src/adaR.cpp [hunk 1]
```diff
@@ -182,21 +188,144 @@ CharacterVector Rcpp_ada_get_host(const CharacterVector& url_vec, bool decode) {
...
 
 // [[Rcpp::export]]
-CharacterVector Rcpp_ada_get_protocol(const CharacterVector& url_vec, bool decode) {
+CharacterVector Rcpp_ada_get_protocol(const CharacterVector& url_vec,
+                                      bool decode) {
   return Rcpp_ada_get(url_vec, &ada_get_protocol, decode);
 }
+
+// higher-order function for bool Rcpp_ada_set_*
+CharacterVector Rcpp_ada_set_bool(
```
Comment (schochastics) [inline src/adaR.cpp:215]: @chainsawriot there are two types of set functions: bool/void, so I defined two Rcpp_ada_set_ main functions. Do you think one can unify them?

Comment (chainsawriot) [reply src/adaR.cpp:215]:
```cpp
template <typename T>
CharacterVector Rcpp_ada_set(
    const CharacterVector& url_vec,
    std::function<T(ada_url, const char*, size_t)> func,
    const CharacterVector& subst, bool decode) {
  unsigned int n = url_vec.length();
  CharacterVector out(n);
  for (int i = 0; i < url_vec.length(); i++) {
    String s = url_vec[i];
    String s2 = subst[i];
    std::string_view input(s.get_cstring());
    std::string_view replace(s2.get_cstring());
    ada_url url = ada_parse(input.data(), input.length());
    if (!ada_is_valid(url)) {
      out[i] = NA_STRING;
    } else {
      func(url, replace.data(), replace.length());
      out[i] = charsub(ada_get_href(url));
    }
    ada_free(url);
  }
  if (decode) {
    out = Rcpp_url_decode2(out);
  }
  return (out);
}

// [[Rcpp::export]]
CharacterVector Rcpp_ada_set_href(const CharacterVector& url_vec,
                                  const CharacterVector& subst, bool decode) {
  return Rcpp_ada_set<bool>(url_vec, &ada_set_href, subst, decode);
}

// [[Rcpp::export]]
CharacterVector Rcpp_ada_set_hash(const CharacterVector& url_vec,
                                  const CharacterVector& subst, bool decode) {
  return Rcpp_ada_set<void>(url_vec, &ada_set_hash, subst, decode);
}

```

Comment (schochastics) [reply src/adaR.cpp:215]: thanks, will integrate that later tonight

**Pull request merged**